KUDO can get stuck in plan when indirect resource is referenced

**What happened**:
We have an operator that uses `cert-manager` to create a Secret which is used by a subsequent deployment:
- Task 1: Apply `Issuer` resource
- Task 2: Apply `Deployment` which references the Secret that will be created by `cert-manager`.

KUDO runs Task 1 and applies the `Issuer` resource, then creates an error and does not continue with the plan execution:
```
           ├── Step certificate [ERROR] (A transient error when executing task deploy.deploy.certificate.certificate. Will retry. failed to add dependencies hash: failed to add dependency hash to istio-system/cacert-job-1.5.1: error calculating hash for istio-ca of type /v1, Kind=Secret: failed to get dependeny istio-system/istio-ca: failed to retrieve object istio-system/istio-ca: secrets "istio-ca" not found)
``` 

Although the error message mentions a Retry, the `last updated` time does not change:
```
Plan deploy (serial strategy) [IN_PROGRESS], last updated 2020-05-13 19:02:44
```

**What you expected to happen**:
KUDO should retry the task when the Secret is created by cert-manager and apply the resource.

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:
This is a problem with the reconciliation:
- KUDO only receives reconciliation requests when a `Watched` resource is Created/Updated/Deleted.
- We do currently not Watch Secrets or ConfigMaps, as they are usually just applied and then are considered healthy.
- In this case we need to watch a resource that is not (directly) created by KUDO, but indirectly by KUDO -> Issuer -> Secret.
- cert-manager obviously doesn't set the KUDO heritage labels, but it should at least set correct ownerReferences that we might be able to follow that would bring us back to the instance
- We could also start temporary watches when we miss a specific resource that is required by the dependency hash calculation.



**Environment**:
- Kubernetes version (use `kubectl version`):
- Kudo version (use `kubectl kudo version`): 
- Operator:
- operatorVersion:
- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

--- a/pkg/engine/renderer/dependencies.go
+++ b/pkg/engine/renderer/dependencies.go
@@ -188,7 +188,7 @@ func sanitizeAndSerialize(origObj *unstructured.Unstructured) (string, error) {
 	obj.SetSelfLink("")
 	obj.SetUID("")
 
-	// This may be set or not, depending if the original object is typed or unstructured. To be safe we nil it
+	// This may be set or not, depending if the original object is typed or unstructured. To be safe we clear it
 	obj.SetGroupVersionKind(schema.GroupVersionKind{})
 
 	// Annotations are notorious for containing quickly changing strings: plan/phase/task names, uids, dates, etc.
@@ -197,9 +197,8 @@ func sanitizeAndSerialize(origObj *unstructured.Unstructured) (string, error) {
 	return ToYaml(obj)
 }
 
-// resourceDependency returns the resource of type t with the given namespace/name, either from the passed in list of
-// objects or the last applied configuration from the API server. If the resource is not found, the func returns
-// nil and no error
+// resourceDependency returns the specified resource, either from the passed in list of objects or the last applied
+// configuration from the API server.
 func (de *dependencyCalculator) resourceDependency(d resourceDependency) (*unstructured.Unstructured, error) {
 
 	// First try to find the dependency in the local list, if it's deployed in the same task we'll find it here

--- a/pkg/engine/renderer/dependencies.go
+++ b/pkg/engine/renderer/dependencies.go
@@ -198,7 +198,7 @@ func sanitizeAndSerialize(origObj *unstructured.Unstructured) (string, error) {
 }
 
 // resourceDependency returns the specified resource, either from the passed in list of objects or the last applied
-// configuration from the API server.
+// configuration from the API server. If the resource is NotFound, the func returns nil and no error
 func (de *dependencyCalculator) resourceDependency(d resourceDependency) (*unstructured.Unstructured, error) {
 
 	// First try to find the dependency in the local list, if it's deployed in the same task we'll find it here
@@ -211,7 +211,6 @@ func (de *dependencyCalculator) resourceDependency(d resourceDependency) (*unstr
 	}
 
 	// We haven't found it, so we need to query the api server to get the current version
-	//dep, _ := reflect.New(t).Interface().(metav1.Object)
 	dep := &unstructured.Unstructured{}
 	dep.SetGroupVersionKind(d.gvk)
 

--- a/pkg/engine/renderer/dependencies.go
+++ b/pkg/engine/renderer/dependencies.go
@@ -4,14 +4,15 @@ import (
 	"context"
 	"crypto/md5" //nolint:gosec
 	"fmt"
+	"log"
 	"sort"
 
-	"k8s.io/apimachinery/pkg/runtime/schema"
-
 	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 
 	"github.com/kudobuilder/kudo/pkg/util/kudo"
@@ -138,16 +139,25 @@ func (de *dependencyCalculator) getHashForDependency(d resourceDependency) (hash
 	if err != nil {
 		return hashBytes{}, fmt.Errorf("failed to get dependeny %s/%s: %v", d.namespace, d.name, err)
 	}
+
+	if dep == nil {
+		// We ignore NotFound resources here. It would be better to fail and wait for a retry, but at the moment
+		// we may not get a reconcile request when the resource is available, as it may not be deployed directly
+		// by KUDO but indirectly (i.e. cert-manager creates a secret that is referenced here)
+		log.Printf("Resource %s/%s was not found for dependency calculation, skipping it", d.namespace, d.name)
+		de.cache[d] = hashBytes{}
+		return de.cache[d], nil
+	}
 	if _, ok := dep.GetAnnotations()[kudo.SkipHashCalculationAnnotation]; ok {
 		de.cache[d] = hashBytes{}
-	} else {
-		yamlStr, err := sanitizeAndSerialize(dep)
-		if err != nil {
-			return hashBytes{}, fmt.Errorf("failed to serialize dependeny %s/%s: %v", d.namespace, d.name, err)
-		}
-		de.cache[d] = md5.Sum([]byte(yamlStr)) //nolint:gosec
+		return de.cache[d], nil
 	}
 
+	yamlStr, err := sanitizeAndSerialize(dep)
+	if err != nil {
+		return hashBytes{}, fmt.Errorf("failed to serialize dependeny %s/%s: %v", d.namespace, d.name, err)
+	}
+	de.cache[d] = md5.Sum([]byte(yamlStr)) //nolint:gosec
 	return de.cache[d], nil
 }
 
@@ -167,7 +177,9 @@ func sanitizeAndSerialize(origObj *unstructured.Unstructured) (string, error) {
 	return ToYaml(obj)
 }
 
-// resourceDependency returns the resource of type t with the given namespace/name, either from the passed in list of objects or the last applied configuration from the API server
+// resourceDependency returns the resource of type t with the given namespace/name, either from the passed in list of
+// objects or the last applied configuration from the API server. If the resource is not found, the func returns
+// nil and no error
 func (de *dependencyCalculator) resourceDependency(d resourceDependency) (*unstructured.Unstructured, error) {
 
 	// First try to find the dependency in the local list, if it's deployed in the same task we'll find it here
@@ -190,6 +202,9 @@ func (de *dependencyCalculator) resourceDependency(d resourceDependency) (*unstr
 	}
 
 	err := de.Client.Get(context.TODO(), key, dep)
+	if apierrors.IsNotFound(err) {
+		return nil, nil
+	}
 	if err != nil {
 		return nil, fmt.Errorf("failed to retrieve object %s/%s: %v", d.namespace, d.name, err)
 	}

--- a/pkg/engine/renderer/dependencies.go
+++ b/pkg/engine/renderer/dependencies.go
@@ -171,6 +171,26 @@ func sanitizeAndSerialize(origObj *unstructured.Unstructured) (string, error) {
 	// OwnerReferences need to be skipped as they contain a changing UID
 	obj.SetOwnerReferences([]metav1.OwnerReference{})
 
+	// These fields should not be present when we get the obj from our task or from the LastAppliedConfigAnnotation, but
+	// we may encounter objects that are not deployed by KUDO and therefore do not have the LastAppliedConfigAnnotation.
+	// This is a best effort attempt to make the hash as stable as possible
+	v := int64(0)
+	obj.SetResourceVersion("")
+	obj.SetClusterName("")
+	obj.SetFinalizers([]string{})
+	obj.SetContinue("")
+	obj.SetCreationTimestamp(metav1.Time{})
+	obj.SetDeletionGracePeriodSeconds(&v)
+	obj.SetGenerateName("")
+	obj.SetGeneration(0)
+	obj.SetManagedFields([]metav1.ManagedFieldsEntry{})
+	obj.SetRemainingItemCount(&v)
+	obj.SetSelfLink("")
+	obj.SetUID("")
+
+	// This may be set or not, depending if the original object is typed or unstructured. To be safe we nil it
+	obj.SetGroupVersionKind(schema.GroupVersionKind{})
+
 	// Annotations are notorious for containing quickly changing strings: plan/phase/task names, uids, dates, etc.
 	obj.SetAnnotations(map[string]string{})
 
@@ -213,7 +233,8 @@ func (de *dependencyCalculator) resourceDependency(d resourceDependency) (*unstr
 	// We use the LastAppliedConfigAnnotation that stores exactly what we applied last time
 	lastConfiguration, ok := dep.GetAnnotations()[kudo.LastAppliedConfigAnnotation]
 	if !ok {
-		return nil, fmt.Errorf("LastAppliedConfigAnnotation is not available on %s/%s", d.namespace, d.name)
+		log.Printf("LastAppliedConfigAnnotation is not available on %s/%s, using resource directly", d.namespace, d.name)
+		return dep, nil
 	}
 
 	obj, err := runtime.Decode(unstructured.UnstructuredJSONScheme, []byte(lastConfiguration))

--- a/pkg/engine/renderer/dependencies_test.go
+++ b/pkg/engine/renderer/dependencies_test.go
@@ -33,6 +33,7 @@ func TestGetResources(t *testing.T) {
 			"key": "value",
 		},
 	}
+	cmWithoutAnnotation := cm.DeepCopy()
 	cmString, _ := runtime.Encode(unstructured.UnstructuredJSONScheme, &cm)
 	cm.Annotations[kudo.LastAppliedConfigAnnotation] = string(cmString)
 
@@ -46,6 +47,7 @@ func TestGetResources(t *testing.T) {
 	}{
 		{name: "from api server", taskObjects: []*unstructured.Unstructured{}, client: fake.NewFakeClientWithScheme(scheme.Scheme, &cm)},
 		{name: "from task objects", taskObjects: []*unstructured.Unstructured{&cmUnstructured}, client: fake.NewFakeClientWithScheme(scheme.Scheme)},
+		{name: "from api server without annotation", taskObjects: []*unstructured.Unstructured{}, client: fake.NewFakeClientWithScheme(scheme.Scheme, cmWithoutAnnotation)},
 	}
 
 	for _, tt := range tests {

--- a/pkg/engine/renderer/enhancer_test.go
+++ b/pkg/engine/renderer/enhancer_test.go
@@ -14,12 +14,12 @@ import (
 	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/apimachinery/pkg/runtime/schema"
 	"k8s.io/apimachinery/pkg/util/uuid"
+	"k8s.io/kubectl/pkg/scheme"
 	clientfake "sigs.k8s.io/controller-runtime/pkg/client/fake"
 	"sigs.k8s.io/yaml"
 
 	"github.com/kudobuilder/kuttl/pkg/test/utils"
 
-	"github.com/kudobuilder/kudo/pkg/client/clientset/versioned/scheme"
 	"github.com/kudobuilder/kudo/pkg/engine"
 	"github.com/kudobuilder/kudo/pkg/test/fake"
 	"github.com/kudobuilder/kudo/pkg/util/kudo"
@@ -234,6 +234,54 @@ func TestEnhancerApply_dependencyHash_unavailableResource(t *testing.T) {
 	assert.NotNil(t, hash, "Pod template spec annotations contains no dependency hash field")
 }
 
+func TestEnhancerApply_dependencyHash_calculatedOnResourceWithoutLastAppliedConfigAnnotation(t *testing.T) {
+	// We may encounter references to resources that are not deployed by KUDO and do not have the
+	// LastAppliedConfigAnnotation. We still need to calculate a mostly stable hash from the resource
+
+	ss := statefulSet("statefulset", "default")
+	cm := configMap("configmap", "default")
+
+	ss.Spec.Template.Spec.Volumes = append(ss.Spec.Template.Spec.Volumes, corev1.Volume{
+		Name: "configMap",
+		VolumeSource: corev1.VolumeSource{
+			ConfigMap: &corev1.ConfigMapVolumeSource{
+				LocalObjectReference: corev1.LocalObjectReference{Name: cm.Name},
+			},
+		},
+	})
+
+	tpls := map[string]string{
+		"statefulset": resourceAsString(ss),
+	}
+
+	meta := metadata()
+	meta.PlanUID = uuid.NewUUID()
+
+	e := &DefaultEnhancer{
+		Scheme:    utils.Scheme(),
+		Discovery: fake.CachedDiscoveryClient(),
+		Client:    clientfake.NewFakeClientWithScheme(scheme.Scheme, cm),
+	}
+
+	objs, err := e.Apply(tpls, meta)
+	if err != nil {
+		t.Errorf("failed to apply template %s", err)
+	}
+
+	ssApplied := funk.Find(objs, func(o runtime.Object) bool {
+		return o.GetObjectKind().GroupVersionKind() == schema.GroupVersionKind{Group: "apps", Version: "v1", Kind: "StatefulSet"}
+	})
+
+	unstructMap, err := runtime.DefaultUnstructuredConverter.ToUnstructured(ssApplied)
+	assert.Nil(t, err, "failed to parse object to unstructured: %s", err)
+
+	annotations, _, _ := unstructured.NestedMap(unstructMap, "spec", "template", "metadata", "annotations")
+	assert.NotNil(t, annotations, "Statefulset pod template spec contains no annotations")
+
+	hash := annotations[kudo.DependenciesHashAnnotation]
+	assert.NotNil(t, hash, "Pod template spec annotations contains no dependency hash field")
+}
+
 func TestEnhancerApply_dependencyHash_changes(t *testing.T) {
 	ss := statefulSet("statefulset", "default")
 	cm := configMap("configmap", "default")
@@ -277,7 +325,6 @@ func TestEnhancerApply_dependencyHash_changes(t *testing.T) {
 
 	hash := annotations[kudo.DependenciesHashAnnotation]
 	assert.NotNil(t, hash, "Pod template spec annotations contains no dependency hash field")
-	assert.Equal(t, "929a2dffa86ad2460fdcf72977998bd0", hash, "Hashes are not the same")
 
 	cm.Data["newkey"] = "newvalue"
 	tpls = map[string]string{

--- a/pkg/engine/renderer/enhancer_test.go
+++ b/pkg/engine/renderer/enhancer_test.go
@@ -14,12 +14,15 @@ import (
 	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/apimachinery/pkg/runtime/schema"
 	"k8s.io/apimachinery/pkg/util/uuid"
+	clientfake "sigs.k8s.io/controller-runtime/pkg/client/fake"
 	"sigs.k8s.io/yaml"
 
+	"github.com/kudobuilder/kuttl/pkg/test/utils"
+
+	"github.com/kudobuilder/kudo/pkg/client/clientset/versioned/scheme"
 	"github.com/kudobuilder/kudo/pkg/engine"
 	"github.com/kudobuilder/kudo/pkg/test/fake"
 	"github.com/kudobuilder/kudo/pkg/util/kudo"
-	"github.com/kudobuilder/kuttl/pkg/test/utils"
 )
 
 func TestEnhancerApply_embeddedMetadataStatefulSet(t *testing.T) {
@@ -183,6 +186,52 @@ func TestEnhancerApply_dependencyHash_noDependencies(t *testing.T) {
 
 	hash := annotations[kudo.DependenciesHashAnnotation]
 	assert.Nil(t, hash, "Pod template spec annotations contains a dependency hash but no dependencies")
+}
+
+func TestEnhancerApply_dependencyHash_unavailableResource(t *testing.T) {
+	// Test that the dependency calculation does not error out on a resource that is NotAvailable at the moment
+
+	ss := statefulSet("statefulset", "default")
+
+	ss.Spec.Template.Spec.Volumes = append(ss.Spec.Template.Spec.Volumes, corev1.Volume{
+		Name: "configMap",
+		VolumeSource: corev1.VolumeSource{
+			ConfigMap: &corev1.ConfigMapVolumeSource{
+				LocalObjectReference: corev1.LocalObjectReference{Name: "unvailableConfigMap"},
+			},
+		},
+	})
+
+	tpls := map[string]string{
+		"statefulset": resourceAsString(ss),
+	}
+
+	meta := metadata()
+	meta.PlanUID = uuid.NewUUID()
+
+	e := &DefaultEnhancer{
+		Client:    clientfake.NewFakeClientWithScheme(scheme.Scheme),
+		Scheme:    utils.Scheme(),
+		Discovery: fake.CachedDiscoveryClient(),
+	}
+
+	objs, err := e.Apply(tpls, meta)
+	if err != nil {
+		t.Errorf("failed to apply template %s", err)
+	}
+
+	ssApplied := funk.Find(objs, func(o runtime.Object) bool {
+		return o.GetObjectKind().GroupVersionKind() == schema.GroupVersionKind{Group: "apps", Version: "v1", Kind: "StatefulSet"}
+	})
+
+	unstructMap, err := runtime.DefaultUnstructuredConverter.ToUnstructured(ssApplied)
+	assert.Nil(t, err, "failed to parse object to unstructured: %s", err)
+
+	annotations, _, _ := unstructured.NestedMap(unstructMap, "spec", "template", "metadata", "annotations")
+	assert.NotNil(t, annotations, "Statefulset pod template spec contains no annotations")
+
+	hash := annotations[kudo.DependenciesHashAnnotation]
+	assert.NotNil(t, hash, "Pod template spec annotations contains no dependency hash field")
 }
 
 func TestEnhancerApply_dependencyHash_changes(t *testing.T) {

--- a/test/e2e/dependencies-hash/01-assert.yaml
+++ b/test/e2e/dependencies-hash/01-assert.yaml
@@ -7,7 +7,7 @@ spec:
   template:
     metadata:
       annotations:
-        kudo.dev/dependencies-hash: af5aeaa4a7dbc49b5d5fd9090541c141
+        kudo.dev/dependencies-hash: 6fa812afe527bdb6c9a5fbfcba8193bb
         kudo.dev/operator-version: 0.1.0
     spec:
       containers:

--- a/test/e2e/dependencies-hash/02-assert.yaml
+++ b/test/e2e/dependencies-hash/02-assert.yaml
@@ -8,7 +8,7 @@ spec:
     metadata:
       annotations:
         # This update to the deployment must have the same hash as 00-assert.yaml
-        kudo.dev/dependencies-hash: af5aeaa4a7dbc49b5d5fd9090541c141
+        kudo.dev/dependencies-hash: 6fa812afe527bdb6c9a5fbfcba8193bb
         kudo.dev/operator-version: 0.1.0
     spec:
       containers:

--- a/test/e2e/dependencies-hash/04-assert.yaml
+++ b/test/e2e/dependencies-hash/04-assert.yaml
@@ -8,7 +8,7 @@ spec:
     metadata:
       annotations:
         # After the second update to replicas the hash should change, based on the last deployed version of the config map
-        kudo.dev/dependencies-hash: 932a6cf74881302904011a9ad09cfe6e
+        kudo.dev/dependencies-hash: da9f56a62d3c4af2596bebace008d845
         kudo.dev/operator-version: 0.1.0
     spec:
       containers:

--- a/test/e2e/dependencies-hash/06-assert.yaml
+++ b/test/e2e/dependencies-hash/06-assert.yaml
@@ -8,7 +8,7 @@ spec:
     metadata:
       annotations:
         # After this config map update the hash should change, as the deployment was applied as well
-        kudo.dev/dependencies-hash: 5cbc3d2ae8bc41ebc23192af530566b6
+        kudo.dev/dependencies-hash: 1510654b540619228b477bdd56d44175
         kudo.dev/operator-version: 0.1.0
     spec:
       containers:

